File: solutions/erlang/all-your-base/1/src/all_your_base.erl
-module(all_your_base).

-export([rebase/3, toBase/2, fromBase/2]).

fromBase(Digits, Base) ->
    lists:foldl(fun(Digit, Acc) -> Acc * Base + Digit end, 0, Digits).

toBase(0, _, []) ->
    [0];
toBase(0, _, Acc) ->
    Acc;
toBase(Value, Base, Acc) ->
    toBase(Value div Base, Base, [Value rem Base | Acc]).

toBase(Digits, Base) ->
    toBase(Digits, Base, []).

rebase(_, InputBase, _) when InputBase < 2 ->
    {error, "input base must be >= 2"};
rebase(_, _, OutputBase) when OutputBase < 2 ->
    {error, "output base must be >= 2"};
rebase(Digits, InputBase, OutputBase) ->
    ValidDigits = lists:all(fun(Digit) -> Digit < InputBase andalso Digit >= 0 end, Digits),
    if ValidDigits ->
           {ok, toBase(fromBase(Digits, InputBase), OutputBase)};
       true ->
           {error, "all digits must satisfy 0 <= d < input base"}
    end.

File: solutions/lua/all-your-base/1/all-your-base.lua
local all_your_base = {}

all_your_base.convert = function(from_digits, from_base)
    if from_base < 2 then
        error('invalid input base')
    end
    local value = 0
    for _, digit in ipairs(from_digits) do
        if digit < 0 then
            error('negative digits are not allowed')
        end
        if digit >= from_base then
            error('digit out of range')
        end
        value = value * from_base
        value = value + digit
    end
    local converter = {}
    converter.to = function(to_base)
        if to_base < 2 then
            error('invalid output base')
        end
        local iter_value = value
        local num_digits = 0
        while iter_value ~= 0 do
            iter_value = iter_value // to_base
            num_digits = num_digits + 1
        end
        iter_value = value
        local result = {}
        for index = 1, num_digits do
            result[num_digits - index + 1] = iter_value % to_base
            iter_value = iter_value // to_base
        end
        if num_digits == 0 then
            result[1] = 0
        end
        return result
    end
    return converter
end

return all_your_base
